### PR TITLE
Fix AJAX data source error

### LINKED ISSUE #4355: AJAX data source "the results could not be loaded" error
## Prerequisites
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of Select2
- [x] This is not a usage question (Those should be directed to the [community](https://select2.github.io/community.html))
- [x] I have attempted to find the simplest possible steos to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)
## Steps to reproduce the issue
1. Navigate to: https://select2.github.io/examples.html#data-ajax
2. Input text
3. Before the ajax results are loaded, input more text
## Expected behavior and actual behavior

When I follow those steps, I see...

"The results could not be loaded." is displayed (then, after a while, the results are populated)

I was expecting...

"Searching.." displayed as more text is inputted
## Environment

Browsers
- [x] Google Chrome
- [x] Mozilla Firefox
- [ ] Internet Explorer

Operating System
- [x] Windows
- [x] Mac OS X
- [x] Linux
- [ ] Mobile

Libraries
- jQuery version: v1.11.2
- Select2 version: master (v4.0.2#master)
## Isolating the problem
- [x] This bug happens [on the examples page](https://select2.github.io/examples.html)
- [x] The bug happens consistently across all tested browsers
- [x] This bug happens when using Select2 without other pluigns
- [x] I can reproduce this bug in [a jsbin](https://jsbin.com/)

### PR DESCRIPTION
**Pull request by MichaelMackus**
This pull request includes a
- [x] Bug fix

The following changes were made
- Check for existence of "status" in the $request object (instead of truthfulness)
- Check that $request.status === 0 (number) not '0' (string)

Fixes #4355

Verified in Firefox & Chromium

### CODE REVIEW

**Comment by PowerKiKi** [comment]
I believe this the solution to a bug that is affecting all installation of select2. As such, shouldn't it be released in a patch version (4.0.4), rather than a minor version (4.1.0) ?

It seems that including this for 4.1 is only delaying the fix without any other benefits. I don't think it's either a breaking change or a new feature.

**Comment by kevin-brown** [comment]
@PowerKiKi the next planned release of Select2 is 4.1.0, we have no intentions of rolling a 4.0.4 release due to (outside) time constraints.

**Comment by PowerKiKi** [comment]
Not quite what I wanted to hear, but thanks for taking the time to answer anyway :+1:

**Comment by drmmr763** [comment]
Currently experiencing bug #4355. Would love to see this PR merged or the next release rolled out so I can upgrade. Probably going to apply this patch anyway for the time being. But +1 this PR!

**Comment by MichaelMackus** [comment]
:+1: for a new minor version increase. This is a bugfix affecting most/all modern browsers!

**Comment by PowerKiKi** [comment]
It seems that select2 lacks a proper workflow for (semi-)continuous delivery. It's a pity because bugs like these impact a lot of users, yet are very simple and safe to fix. Especially when the community do half the work to fix it.

**Comment by vladkazantsev** [comment]
Any comment on when this will be released? Time frames? Experiencing the 4355 across all devices...

File: dist/js/select2.full.js [hunk 1]
```diff
@@ -3444,7 +3444,7 @@ S2.define('select2/data/ajax',[
       }, function () {
         // Attempt to detect if a request was aborted
         // Only works if the transport exposes a status property
-        if ($request.status && $request.status === '0') {
+        if ('status' in $request && $request.status === 0) {
```
Comment (jpic) [inline dist/js/select2.full.js]: Is this not going to break support for all browsers that have '0' ? shouldnt this be `$request.status === 0 || $request.status === '0'`?

File: dist/js/select2.js [hunk 2]
```diff
@@ -3444,7 +3444,7 @@ S2.define('select2/data/ajax',[
       }, function () {
         // Attempt to detect if a request was aborted
         // Only works if the transport exposes a status property
-        if ($request.status && $request.status === '0') {
+        if ('status' in $request && $request.status === 0) {
```
Comment (jpic) [inline dist/js/select2.js]: Don't change dist files, for more info check CONTRIBUTING

File: src/js/select2/data/ajax.js [hunk 3]
```diff
@@ -82,7 +82,7 @@ define([
       }, function () {
         // Attempt to detect if a request was aborted
         // Only works if the transport exposes a status property
-        if ($request.status && $request.status === '0') {
+        if ('status' in $request && $request.status === 0) {
```
Comment (jpic) [inline src/js/select2/data/ajax.js]: Same question.

**Comment by jpic** [comment]
We'll need regression tests before we can merge this.

**Comment by MichaelMackus** [comment]
@jpic done & submitted changes. I'm not familiar with JS testing, but will look into if I find some free time. If somebody else could take on the tests I'd be grateful.

**Comment by alexweissman** [comment]
Seems like a pretty popular fix - can someone who knows their way around the testing framework please write some tests so we can get this released in the next revision?

**Comment by lordplagus02** [comment]
Just reminding everyone that this exists and it's been over a year

**Comment by alexweissman** [comment]
@lordplagus02 it will be released in 4.0.6.  Hold yer horses 🐴

**Comment by alexweissman** [comment]
Merged into `develop` for 4.0.6 release.

File: dist/js/select2.full.js (hunk 1 above)
Comment (M4cker) [reply dist/js/select2.full.js]: For those who are using timeout : if ('status' in $request && $request.status === 0 && $request.statusText != "timeout")

**Comment by MichielDP** [comment]
This bug is **STILL** happening in version 4.0.6-rc.1. I'm getting the error "the results could not be loaded" with even the most basic usage of select2. The request is not even reaching my backend method. When I paste the exact same url in the browser url bar, then it does reach my backend.

**Comment by markoheijnen** [comment]
Any idea when this is going to be released.

**Comment by pedrofurtado** [comment]
@markoheijnen In next weeks. I am reviewing some another PRs. After it, we will release a version.

**Comment by smuszel** [comment]
Hey. Are you sure this solution indeed fixes the issue. At my company we are still having issues with it. Changing function at 3589 of select.js to below fixed the problem.

```
function(_, statusText) {
    if (statusText !== 'abort') {
        self.trigger('results:message', {
            message: 'errorLoading',
        });
    }
},
```

**Pull request closed without merging**